### ISSUE TITLE
Class-based keys ... how should they work?

### ISSUE THREAD

**Issue by jasonrhodes** [bendrucker/fill-keys#6]
Right now you can see from thlorenz/proxyquire#136 that if you use a class-based stub, proxyquire won't "see" those instance methods and will overwrite them with the call-through library that was actually required in, unless you use the noCallThru features of proxyquire.

Tracking that down, it's because fill-keys uses the "merge-descriptors" library which in turn, when told not to replace everything as fill-keys indicates, uses `hasOwnProperty` to check for the existence of a method before setting it. That doesn't work for instance methods of an ES6 class.

I'd be happy to submit a PR to change this but I'm not really familiar enough with descriptors to know if that descriptor merge functionality is important to fill-keys users—I assume that it would be bad to replace that merge altogether with a `for in` type loop similar to the es3 fill-keys function.

In other words:

```javascript
const fill = require('fill-keys')

class Test {
  constructor() {}
  hello() {
    console.log('from inside the class')
  }
}

const dest = new Test()

const src = {
  hello: function() {
    console.log('from the object literal')
  }
}

fill(dest, src)

dest.hello() // => what do you expect to log here? right now you'll get 'from the object literal'
```

If the merge descriptors functionality needs to stay, then I'm not sure how you avoid this. Maybe a separate function exported as `fillKeysClasses` or similar, that skips the merge-descriptors merge and does a `for in` loop instead? Curious to hear thoughts from @bendrucker before I do any PR work. Thanks!

**Comment by bendrucker** [comment]
Was having trouble remembering the thinking behind avoiding a simple `in` lookup but this helped:

https://github.com/thlorenz/proxyquire/commit/3bc3c875a351ecd8ec99537f7110e7c4508a14d2

Prototype chains can get complicated and to keep things sane they need to be copied to match the source. Copying instance methods and just tacking them on via normal assignment would not be okay as far as I'm concerned. 

In order to actually get proper descriptor behavior you'd need to resort to hacks. You'd have to check whether each value is an own property or inherited. If it's inherited you'd have to call `Object.getPrototypeOf` and hope it's owned by the immediate parent prototype. If it's not, you might be out of luck. Once you're at the top you could call `Object.getOwnPropertyDescriptor` to get the value for re-description.

Honestly I think it's better to document this limitation than try to work around it. There were lots of things wrong with the code the OP in https://github.com/thlorenz/proxyquire/issues/136 posted. I don't see Thorsten's example (https://github.com/thlorenz/proxyquire/issues/136#issuecomment-249269877) of stubbing the path module with a class instance as worth supporting since to me there's no logical reason to try to replace stub a plain object with an instance. 

If you can share an example that you see as a legitimate use of an instance as a stub I'm happy to reconsider.

**Comment by jasonrhodes** [comment]
Our use case was, in a test file:

``` javascript
const proxyquire = require('proxyquire')
const ConfigMocker = require('../mocks/ConfigMocker')
const testConfig = {
  some: 'config',
  that: 'you want to mock'
}
const mockConfig = new ConfigMocker(testConfig)
let testModel

beforeEach(function() {
  testModel = proxyquire('../../lib/testModel', {  `@scope/our-config-lib`: mockConfig })
}

it('should do something', function() {
  return testModel.create().then( ... )
  // etc
})

// in the model code
model.create = function() {
  mockConfig.get('some')
  // etc
}
```

Granted we can probably achieve this with reconfiguring our ConfigMocker to not be a class, but it doesn't sound like a totally bizarre case. That said, I get that you can't swap out those descriptors. To be honest this, among other things, makes me wish that `noCallThru` mode was the default for proxyquire. Thanks for the input—feel free to close as far as I'm concerned.

**Comment by bendrucker** [comment]
Sounds good, thanks for the input!

**Comment by bendrucker** [comment]
`noCallThru` depends on the use case. If you want to override something like `http.get` you'd prefer not having every other module that depends on `http` to break by default. For testing configuration strict mocking w/out call-thru is definitely the right choice.
